sql: correctly escape SQL strings

madelynnblue opened the pull request:
Similar to changes in encodeSQLBytes, remove all non-printable characters
from the output. Unlike encodeSQLBytes, this function is unicode-aware
and escapes unicode code points.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7369)

<!-- Reviewable:end -->

RaduBerinde commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/parser/encode.go, line 64 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKomuGGkvJ3lKYCpdBZ:-KKomuGHZCBDERhYKvky:-2137405068) ([raw file](https://github.com/cockroachdb/cockroach/blob/81b1aad9e1df43b9b3945dd0e33c6216e8d8365c/sql/parser/encode.go#L64)):*
> ```Go
> 			nextStart = false
> 		}
> 		if r == utf8.RuneError {
> ```

AFAICT the only case where we don't first call `writeStart` is when we have a printable char. Couldn't we start with a special `if r >= 0x20 && r < 0x80 { continue; }` case then do what `writeStart` does, and then go through the cases?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/parser/encode.go, line 64 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKomuGGkvJ3lKYCpdBZ:-KKonmR8L2JVDDUzzpWx:1380687172) ([raw file](https://github.com/cockroachdb/cockroach/blob/81b1aad9e1df43b9b3945dd0e33c6216e8d8365c/sql/parser/encode.go#L64)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> AFAICT the only case where we don't first call `writeStart` is when we have a printable char. Couldn't we start with a special `if r >= 0x20 && r < 0x80 { continue; }` case then do what `writeStart` does, and then go through the cases?
</details>
Er missed that we have a printable escape, so `if r >= 0x20 && r < 0x80 && encodeMap[ch] == dontEscape`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369)*
<!-- Sent from Reviewable.io -->

knz commented:
Wait what? This runs counter to #7045  and #7048. What do you plan to do about the CLI?
It seems to me that this warrants a boolean flag about whether to escape or not.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-:-KKopIpje2Wn576_KsJt:899045965)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/parser/parse_test.go, line 1148 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKozNQTQsiJBL9TXnNB:-KKozNQTQsiJBL9TXnNC:1737639651) ([raw file](https://github.com/cockroachdb/cockroach/blob/81b1aad9e1df43b9b3945dd0e33c6216e8d8365c/sql/parser/parse_test.go#L1148)):*
> ```Go
> 			s := string([]byte{byte(i), byte(j)})
> 			var buf bytes.Buffer
> 			encodeSQLString(&buf, s)
> ```

did you mean to use `encode` here?

---

*[sql/parser/parse_test.go, line 1158 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKozJlj9aUy1ExWxZT_:-KKozJlj9aUy1ExWxZTa:-148703175) ([raw file](https://github.com/cockroachdb/cockroach/blob/81b1aad9e1df43b9b3945dd0e33c6216e8d8365c/sql/parser/parse_test.go#L1158)):*
> ```Go
> 			stmts, err := parseTraditional(sql)
> 			if err != nil {
> 				t.Fatalf("%s: expected success, but found %s", sql, err)
> ```

why these?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Unicode characters are escaped, like: `\u2603` (that is, those 6 ASCII characters appear), so the CLI is fine. The output of this function won't produce any bytes < 0x20 or >= 0x7F, just like encode bytes. The only difference is this function can print \uXXXX and \UXXXXXXXX.

---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[sql/parser/encode.go, line 64 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKomuGGkvJ3lKYCpdBZ:-KKpj88WsX21BXpEusvu:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/81b1aad9e1df43b9b3945dd0e33c6216e8d8365c/sql/parser/encode.go#L64)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Er missed that we have a printable escape, so `if r >= 0x20 && r < 0x80 && encodeMap[ch] == dontEscape`
</details>
Done.

---

*[sql/parser/parse_test.go, line 1148 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKozNQTQsiJBL9TXnNB:-KKphergPzouTYeS2hhB:-573343871) ([raw file](https://github.com/cockroachdb/cockroach/blob/81b1aad9e1df43b9b3945dd0e33c6216e8d8365c/sql/parser/parse_test.go#L1148)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> did you mean to use `encode` here?
</details>
Oops. Yes.

---

*[sql/parser/parse_test.go, line 1158 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKozJlj9aUy1ExWxZT_:-KKphs7B-56EdBTNPGzr:-853858286) ([raw file](https://github.com/cockroachdb/cockroach/blob/81b1aad9e1df43b9b3945dd0e33c6216e8d8365c/sql/parser/parse_test.go#L1158)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why these?
</details>
To make sure we are producing literals that are parseable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-:-KKopoLyfAUHwrv9JWPp:-781659744)*
<!-- Sent from Reviewable.io -->

knz commented:
Thanks for clarifying.
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r1.
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-:-KKr6PfIjinUPaK-dHbg:1717823825)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[sql/parser/encode.go, line 46 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKrEkIFFzVjonpzu2oF:-KKrEkIFFzVjonpzu2oG:660862632) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0eecf57a28de89f6ab9f18a01831f0990dfaa53/sql/parser/encode.go#L46)):*
> ```Go
> 	start := 0
> 	nextStart := false
> 	var i int
> ```

why are `i` and `r` `var`s rather than `for i, r := range in`?

---

*[sql/parser/encode.go, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKrFRHWcSMr-ROGcQ6v:-KKrFRHXlUMHcV4yYNML:-1398366548) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0eecf57a28de89f6ab9f18a01831f0990dfaa53/sql/parser/encode.go#L63)):*
> ```Go
> 		}
> 		buf.WriteString(in[start:i])
> 		// Since we don't know the byte width of the current unicode code point
> ```

https://godoc.org/unicode/utf8#RuneLen might permit you to get rid of nextStart?

---

*[sql/parser/encode.go, line 72 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKrFpqxfYDSUcNAMfML:-KKrFpqxfYDSUcNAMfMM:1293079732) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0eecf57a28de89f6ab9f18a01831f0990dfaa53/sql/parser/encode.go#L72)):*
> ```Go
> 			// that it's a byte in the case of an error).
> 			buf.Write(hexMap[in[i]])
> 		} else if r < 256 {
> ```

this appears to also want to use https://godoc.org/unicode/utf8#RuneLen

---

*[sql/parser/parse_test.go, line 1158 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKozJlj9aUy1ExWxZT_:-KKrEVfdVrxTfJn5whhx:-94473533) ([raw file](https://github.com/cockroachdb/cockroach/blob/81b1aad9e1df43b9b3945dd0e33c6216e8d8365c/sql/parser/parse_test.go#L1158)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> To make sure we are producing literals that are parseable.
</details>
I meant why these changes from Errorf to Fatalf?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[sql/parser/encode.go, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKt4Kz6YUw6q_jx31bT:-KKt4Kz6YUw6q_jx31bU:1309190934) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0eecf57a28de89f6ab9f18a01831f0990dfaa53/sql/parser/encode.go#L55)):*
> ```Go
> 		}
> 		ch := byte(r)
> 		if r >= 0x20 && r < 0x7F && encodeMap[ch] == dontEscape {
> ```

<= (or < 0x80)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 1 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/parser/encode.go, line 46 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKrEkIFFzVjonpzu2oF:-KKt8ajLIFc7V4vVFTZQ:252503418) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0eecf57a28de89f6ab9f18a01831f0990dfaa53/sql/parser/encode.go#L46)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why are `i` and `r` `var`s rather than `for i, r := range in`?
</details>
Remnant of a previous revision. Fixed.

---

*[sql/parser/encode.go, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKt4Kz6YUw6q_jx31bT:-KKt8mJ8yLbrlPXtEY4Q:950039518) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0eecf57a28de89f6ab9f18a01831f0990dfaa53/sql/parser/encode.go#L55)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> <= (or < 0x80)
</details>
I think `< 0x7F` is correct. See below where we use `>= 0x7F` as the non-printable range. Thus `<` that is printable.

---

*[sql/parser/encode.go, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKrFRHWcSMr-ROGcQ6v:-KKt9pq1ySwDCwXuv-KM:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0eecf57a28de89f6ab9f18a01831f0990dfaa53/sql/parser/encode.go#L63)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> https://godoc.org/unicode/utf8#RuneLen might permit you to get rid of nextStart?
</details>
Done.

---

*[sql/parser/encode.go, line 72 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKrFpqxfYDSUcNAMfML:-KKt9qAR-AFqOEKmiIgf:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0eecf57a28de89f6ab9f18a01831f0990dfaa53/sql/parser/encode.go#L72)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this appears to also want to use https://godoc.org/unicode/utf8#RuneLen
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 1 of 2 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[sql/parser/parse_test.go, line 1158 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369#-KKozJlj9aUy1ExWxZT_:-KKtA0dnllBMwEBEHYnV:337695640) ([raw file](https://github.com/cockroachdb/cockroach/blob/81b1aad9e1df43b9b3945dd0e33c6216e8d8365c/sql/parser/parse_test.go#L1158)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I meant why these changes from Errorf to Fatalf?
</details>
In practice either none would error or most of them. Using Fatalf was more productive because it didn't fill up the screen with thousands of errors.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7369)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: correctly escape SQL strings

Similar to changes in encodeSQLBytes, remove all non-printable characters
from the output. Unlike encodeSQLBytes, this function is unicode-aware
and escapes unicode code points.
